native/src/main/java/io/ballerina/lib/data/yaml/utils/JsonTraverse.java

```diff
@@ -0,0 +1,323 @@
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.ballerina.lib.data.yaml.utils;
+
+import io.ballerina.lib.data.yaml.common.Types;
+import io.ballerina.lib.data.yaml.parser.ParserUtils;
+import io.ballerina.lib.data.yaml.parser.Values;
+import io.ballerina.runtime.api.PredefinedTypes;
+import io.ballerina.runtime.api.TypeTags;
+import io.ballerina.runtime.api.creators.ValueCreator;
+import io.ballerina.runtime.api.flags.SymbolFlags;
+import io.ballerina.runtime.api.types.ArrayType;
+import io.ballerina.runtime.api.types.Field;
+import io.ballerina.runtime.api.types.IntersectionType;
+import io.ballerina.runtime.api.types.MapType;
+import io.ballerina.runtime.api.types.RecordType;
+import io.ballerina.runtime.api.types.TupleType;
+import io.ballerina.runtime.api.types.Type;
+import io.ballerina.runtime.api.types.UnionType;
+import io.ballerina.runtime.api.utils.StringUtils;
+import io.ballerina.runtime.api.utils.TypeUtils;
+import io.ballerina.runtime.api.utils.ValueUtils;
+import io.ballerina.runtime.api.values.BArray;
+import io.ballerina.runtime.api.values.BError;
+import io.ballerina.runtime.api.values.BMap;
+import io.ballerina.runtime.api.values.BString;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Stack;
+
+/**
+ * Traverse json tree.
+ *
+ * @since 0.1.0
+ */
+public class JsonTraverse {
+
+    private static final ThreadLocal<JsonTree> tlJsonTree = ThreadLocal.withInitial(JsonTree::new);
+
+    public static Object traverse(Object json, BMap<BString, Object> options, Type type, Types.YAMLSchema schema) {
+        JsonTree jsonTree = tlJsonTree.get();
+        try {
+            Object allowDataProjection = options.get(Constants.ALLOW_DATA_PROJECTION);
+            jsonTree.schema = schema;
+            if (allowDataProjection instanceof Boolean) {
+                jsonTree.allowDataProjection = false;
+            } else if (allowDataProjection instanceof BMap<?, ?>) {
+                jsonTree.allowDataProjection = true;
+                jsonTree.absentAsNilableType =
+                        (Boolean) ((BMap<?, ?>) allowDataProjection).get(Constants.ABSENT_AS_NILABLE_TYPE);
+                jsonTree.nilAsOptionalField =
+                        (Boolean) ((BMap<?, ?>) allowDataProjection).get(Constants.NIL_AS_OPTIONAL_FIELD);
+            }
+            return jsonTree.traverseJson(json, type);
+        } finally {
+            jsonTree.reset();
+        }
+    }
+
+    private static class JsonTree {
+        Field currentField;
+        Stack<Map<String, Field>> fieldHierarchy = new Stack<>();
+        Stack<Type> restType = new Stack<>();
+        Deque<String> fieldNames = new ArrayDeque<>();
+        Type rootArray;
+        boolean allowDataProjection = true;
+        boolean nilAsOptionalField = false;
```

Comment:
I hope these fields should be true in default

Reply:
No, this are special scenario where user wants to identify when there is nil value the field could be optional. We made the same decision for [jsondata](https://github.com/ballerina-platform/module-ballerina-data.jsondata/blob/d894edaf3fe826eaeceb31a6d8130d7219602cac/ballerina/json_api.bal#L78) as well.